### PR TITLE
Use tracing-based subscriber logging

### PR DESCRIPTION
**Pull request by sorpaas**
This PR changes the log producer from `env_logger` to `tracing-subscriber::fmt::FmtSubscriber`. We have an existing `ProfilingSubscriber`. It is changed to `ProfilingLayer` in this PR, which can be combined with the `FmtSubscriber`. `LogTracer` is used to convert all `log` records to `tracing` events, so any code depending on `log` will still work.

I have not yet changed local `log` dependencies to `tracing`, as I believe it might bring a lot of merge conflicts and may be better done in a separate PR.

rel #6809

polkadot companion: https://github.com/paritytech/polkadot/pull/1552

### CODE REVIEW

**Comment by davidbarsky** [comment]
(Hi, I'm a `tracing` maintainer. I saw this issue.)

You might not need to directly depend on `tracing-log` as calling [`.init()` on the subscriber builder](https://docs.rs/tracing-subscriber/0.2.10/tracing_subscriber/fmt/struct.SubscriberBuilder.html#method.init) will read `log`-emitted events.

File: bin/node/bench/Cargo.toml
Comment (ordian) [inline bin/node/bench/Cargo.toml:41]: Wonder why you removed `default-features = false` of `parity-util-mem`? Seems to be not required for this PR. The thing I'm worried about is feature leaking into other places because of how cargo resolves features in a workspace.

File: client/tracing/src/lib.rs
Comment (mattrutherford) [inline client/tracing/src/lib.rs:260]: The way this is implemented means that this layer will filter globally (inc from the FmtSubscriber), so I think we need to leave it at default, which is always `true`.

Then we need to do the filtering in `new_span` and keep corresponding noop in `on_enter` etc if span `Id` is not found.

Also somehow we need to enable the callsites required for profiling in the FmtSubscriber.

**Comment by mattrutherford** [comment]
Also, currently when we are using the `LogTraceHandler` there is no output, maybe because there is no logger initialised.

**Comment by bkchr** [comment]
What is the status of this?

**Review by gnunicorn — COMMENTED** [review]
Need to remove the `enabled` but looks good otherwise.

It isn't quite clear to me where/how actually tracking happens after this, and whether that implements `clone_span` correctly. Further more, the way that `wasm`-targets work isn't correctly implemented upon all layers after this – in particular for the env-formatter. Something to keep in mind for #6916 .

**Review by gnunicorn — APPROVED** [review]
sorry, wrong checkbox

**Review by dvdplm — APPROVED** [review]
lgtm modulo tiny nits

File: client/cli/src/lib.rs
Comment (bkchr) [inline client/cli/src/lib.rs:240]:
```suggestion
	if tracing_log::LogTracer::init().is_err() {
		log::debug!(
			"💬 Not registering Substrate logger, as there is already a global logger registered!"
		);
	}
```

We should not print this as a `info` log line.

File: client/cli/src/lib.rs
Comment (bkchr) [inline client/cli/src/lib.rs:264]: Could just be an `if let`

File: client/cli/src/lib.rs
Comment (bkchr) [inline client/cli/src/lib.rs:276]: Same

File: client/cli/src/lib.rs
Comment (bkchr) [inline client/cli/src/lib.rs:298]:
```suggestion
		if tracing::subscriber::set_global_default(subscriber.with(profiling)).is_err() {
			tracing::debug!(
				"💬 Not registering Substrate subscriber, as there is already a global subscriber registered!"
			);
		}
```

File: client/cli/src/lib.rs
Comment (bkchr) [inline client/cli/src/lib.rs:305]:
```suggestion
	} else if tracing::subscriber::set_global_default(subscriber).is_err() {
		tracing::debug!(
				"💬 Not registering Substrate subscriber, as there is already a global subscriber registered!"
		);
```

File: primitives/tracing/src/lib.rs
Comment (bkchr) [inline primitives/tracing/src/lib.rs:49]:
```suggestion
/// Try to init a simple tracing subscriber with log compatibility layer.
/// Ignores any error. Useful for testing.
#[cfg(feature = "std")]
```

File: client/cli/src/lib.rs
Comment (bkchr) [inline client/cli/src/lib.rs:233]:
```suggestion
	tracing_targets: Option<String>,
```

**Review by bkchr — COMMENTED** [review]
Some "styling" issues.

We should not print these messages that something could not be registered as `info!` or `warning!`, because otherwise we will spam parachains :P

File: client/cli/src/lib.rs
Comment (dvdplm) [reply client/cli/src/lib.rs:264]: Hmm. Like it's written it's very clear why we're not handling the error.

Comment (bkchr) [reply client/cli/src/lib.rs:264]: The comment can also just be moved above the `if let`

**Comment by sorpaas** [comment]
(PR is ready for review.)

File: client/cli/src/lib.rs
Comment (ordian) [inline client/cli/src/lib.rs:295]: Previously it was an `info!` log, now it's either `debug!` or `warn!`. If this is intended, could add some comments explaining why.

**Review by ordian — APPROVED** [review]
Looks good, just one question.

File: client/cli/src/lib.rs
Comment (sorpaas) [reply client/cli/src/lib.rs:295]: @bkchr mentioned that it will spam parachains otherwise. https://github.com/paritytech/substrate/pull/6825#pullrequestreview-485791529

**Comment by dvdplm** [comment]
@mattrutherford are your concerns still valid?

File: client/cli/src/lib.rs
Comment (bkchr) [reply client/cli/src/lib.rs:295]: Yeah, we spawn two nodes in Cumulus and as we mostly use the default "way" of initializing the nodes we will always end in a situation where the logger is already registered.

Comment (gnunicorn) [reply client/cli/src/lib.rs:295]: looking at it again, I wonder: why not escalate the error? Here and from `init`(which is already `Result`). And for cumulus I wonder whether we should then refactor the code to be usable without calling `init` (which states in its docs that it "must be done only once."). Which from a brief look at it, shouldn't be too hard... like, if `init` was called once (it doesn't even really need the `&self`) there is no need to call it again... and in cumulus cases we would then only call it once...

**Review by bkchr — APPROVED** [review]

**Comment by gnunicorn** [comment]
bot merge

**Pull request merged**
